Increase test coverage: Array.prototype.splice

matedabis opened the pull request:
Branch coverage:
   Before: 52/72
   After: 72/72

Also found an opportunity to optimize, that "else if" branch is not needed,
should be replaced with an "else".

JerryScript-DCO-1.0-Signed-off-by: Mate Dabis mdabis@inf.u-szeged.hu

LaszloLango commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,174 @@ try {
+  assert (e instanceof ReferenceError);
+}
+
+try{
```
Missing space before `'{'`
There are many minor style issues like this in the added test file. Please always put a space before `'{'`, put semicolon after statements, do not write space before `'('` in function calls (e.g.: `assert(false);`, etc.)

Check your code and fix these issues.

matedabis commented:
@LaszloLango
Corrected the mistakes mentioned.

LaszloLango commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,174 @@ try {
+try {
+  Array.prototype.splice.call(undefined);
+  assert(false);
+} catch(e) {
```
Missing space: `} catch (e) {`
Please fix it in the whole file.

matedabis commented:
Fixed

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,174 @@ try {
+  obj = {get: f, valueOf : f, toString: f};
+  arr = [1, 2, obj, 4, 5];
+  Object.defineProperty(arr, '2',{ 'get' : f } );
+  print(arr);
```
`print` ??? Why is it even needed?

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,174 @@ try {
+  var a = [1, 5, 6, 7, 8, 5];
+  Object.defineProperty(a, '1', { 'get' : function() { throw new ReferenceError("foo"); } });
+  Object.defineProperty(a, '0', { 'get' : function() { throw new ReferenceError("foo"); } });
+  Object.defineProperty(a, '2', { 'get' : function() { throw new ReferenceError("foo"); } });
```
Use different error messages, and assert in the catch block which error has been thrown.

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,174 @@ try {
+  Array.prototype.splice.call(a, function() { });
+  assert(false);
+} catch (e) {
+  assert(e instanceof ReferenceError);
```
Check error messages as well!

rerobika commented:
Also I'd rather call the `else if` to `else` replacement to optimization than mistake correction in the commit message.

matedabis commented:
@rerobika 
That `print` is needed to make an `ECMA_TRY_CATCH` fail and acces that branch in line 1346.
I corrected the other issues.

matedabis commented:
Corrected.

matedabis commented:
@akosthekiss 
The function we are testing: `ecma_builtin_array_prototype_object_splice`
The test file: `/tests/jerry/array-prototype-splice.js`
I dont know yet the result if all test suites are executed. By all test suites of the project do you mean all the tests in `run-tests.py` or all the `.js` files in the `tests` folder, or both?

akosthekiss commented:
I almost forgot about unit tests and various configurations. So, I think combining the coverage results of all run-tests.py executions is the best way.

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,203 @@ try {
+}
+
+var a = {length : 13};
+Array.prototype.splice.call(a, function () {delete a});
```
Spaces around the function body's expression statement.

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,203 @@ try {
+} catch (e) {}
+
+try {
+  f = function() { for(var i = 0; i < arr.length; i++) { delete arr[i]} };
```
Break it into multiple lines.

akosthekiss replied:
+1. (Plus: there are multiple similar constructs in the code; apply a consistent style everywhere)

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,203 @@ try {
+  Object.defineProperty(arr, '2', { 'get' : f });
+  Object.defineProperty(arr, '3', { 'get' : f });
+  Object.defineProperty(arr, '4', { 'get' : f });
+  Object.defineProperty(arr, '5', { 'get' : f });
```
How about using a for loop for these 6 property definitions?

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,203 @@ try {
+  assert(e.message == "4");
+}
+
+try{
```
Could you please add a short description for each new testcase? After this point the test cases are slightly complicated, it would be good to know what is happening.

matedabis commented:
I updated the patch according to the review.

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+  assert(e instanceof TypeError);
+}
+
+// Checking behaior when elements are getting deleted the same way, but there are more elements and arguments
```
typo: behaior

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+  Object.defineProperty(arr, '8', { 'get' : f });
+  Object.defineProperty(arr, '9', { 'get' : f2 });
+  Object.defineProperty(arr, '10', { 'get' : f3 });
+  arr.splice(1, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5);
```
Is this really the simplest use case to test a feature of the engine? Is a 24 element array, 7 manually defined properties, and a 21 element splice needed to cover some path?

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+  assert(e instanceof TypeError);
+}
+
+// Checking behavior when first 3 elements throw error
```
Which code path needs the first/last _3_ elements to be handled specially?

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+
+// Checking behavior when last 3 elements throw error
+try {
+  f = function() {  for(var i=0; i<arr.length; i++) { delete arr[i]} };
```
style: spacing (too many + missing). also: missing ; from after delete statement.

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+
+// Checking behavior when the last element gets deleted
+try {
+  f = function () { delete arr[23];};
```
style: spacing

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+  f = function () { delete arr[23];};
+  arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24];
+  delete arr[23];
+  Object.defineProperty(arr, '23',{ 'get' : f });
```
style: spacing

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+  f = function() { for(var i = 0; i < arr.length; i++) { delete arr[i]} };
+  f1 = function() { delete arr[2];};
+  f2 = function() { delete arr[3];};
+  f3 = function() { delete arr[4];};
```
style: spacing

akosthekiss commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,221 @@ try {
+
+// Checking behaior when elements are getting deleted the same way, but there are more elements and arguments
+try {
+  f = function() { for(var i=0; i<arr.length; i++) { delete arr[i]} };
```
style: spacing, semicolon

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,228 @@ try {
+        delete arr[3];
+        arr.length = 13;
+        Object.defineProperty(arr, '5', function() { });
+      };
```
This style has became even worse than it was before. Please use the following format:
``` js
f = function() {
  delete arr[3];
  arr.length = 13;
  Object.defineProperty(arr, '5', function() { });
};
```

rerobika commented on tests/jerry/array-prototype-splice.js:
```diff
@@ -175,3 +175,228 @@ try {
+  arr = [1, 2, obj, 4, 5];
+  Object.defineProperty(arr, '2',{ 'get' : f } );
+  for(var i = 0; i < arr.length; i++)
+  {
```
Move the open curly brace up:
```js
for (var i = 0; i < arr.length; i++) {
//...body 
}

matedabis commented:
Updated the patch:
- Added comments for each case containing which part of the ES6 it tests.
- Sorted new tests in ascending order
- Corrected the patch according to the reviews. I made some test cases simpler

matedabis commented:
Updated the patch:
- Added comments for each case containing which part of the ES 5.1 it tests.
- Sorted new tests in ascending order
- Corrected the patch according to the reviews. I made some test cases simpler
- Ran the test code in Gecko, V8, SpiderMonkey engines and got the same result.

matedabis commented:
Summary:
The following script is made for testing branch coverage with all the test suites (--jerry-test-suite --test262 --unittests --jerry-tests), or with only one .js file.

https://github.com/matedabis/jerryscript/blob/gcov_coverage_tester/tests/gcov-tests/gcovtester.py

While measuring the branch coverage we don't count JERRY_ASSERT s.

Branch coverage (covered branch/all branch):
Before: 56/78
After: 78/78

Also found an opportunity to optimize, that "else if" branch is not needed,
should be replaced with an "else".

matedabis commented:
Closing PR because of moving into a unified PR.
